docs/source/caching/programmatic-cache-keys.mdx

```diff
@@ -0,0 +1,188 @@
+---
+title: Programmatic Cache Keys
+---
+
+In situations where [declarative cache IDs](./cache-key-resolution) don't fit your use case, you can programmatically generate cache IDs for object types in your normalized cache.
+
+To configure cache keys programmatically, you can alter the [`SchemaConfiguration.cacheKeyInfo(for type:object:)`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/schemaconfiguration/cachekeyinfo(for:object:)) method to return a `CacheKeyInfo` object.
+
+> **Note:** Programmatic cache key usage will take precedence over declarative cache keys. If you return a `CacheKeyInfo` for an object, it will be used to compute that object's cache key and any relevant `@typePolicy` will be ignored. Returning `nil` from this function will still fall back to using declarative cache keys.
+
+## `CacheKeyInfo`
+
+The information needed to construct a **cache key** is represented by a [`CacheKeyInfo`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/cachekeyinfo) value. This `struct` consists of two properties you can provide to affect how a cache key is computed:
+
+1. `let uniqueKeyGroup: String?`
+
+    An optional **group identifier** for a set of objects that should be grouped together in the normalized cache. This is used as the first component of the **cache key**.
+
+    > #### Important: Cache key group uniqueness
+    >
+    > All objects with the same `uniqueKeyGroup` must have unique `id`s across all types.
+
+    To prevent cache key collisions, cache keys will always have a **group identifier** component. When the `uniqueKeyGroup` is `nil` (the default value), to [`__typename`](https://spec.graphql.org/draft/#sec-Type-Name-Introspection) of the response object is used as the **group identifier** by default.
```

Comment:
```suggestion
    To prevent cache key collisions, cache keys will always have a **group identifier** component. When the `uniqueKeyGroup` is `nil` (the default value), the [`__typename`](https://spec.graphql.org/draft/#sec-Type-Name-Introspection) of the response object is used as the **group identifier**.
```